Handle gRPC `@Blocking` in superclasses as well

snazy opened the pull request:
Updates the "blocking methods" discovery mechanism in `io.quarkus.grpc.deployment.GrpcServerProcessor`.

Two changes:
* Discover all gRPC methods from the base class that implements `io.grpc.BindableService` and use the names of the discovered methods to find blocking methods.
* Discover whether gRPC methods shall be blocking by inspecting the `@Blocking`/`@NonBlocking`/`@Transactional` annotations on method and class level including base classes.

Whether a gRPC method is blocking is determined using the following rules, first matching rule applies:
1. blocking, if the top-most method override has a io.smallrye.common.annotation.Blocking annotation.
1. not-blocking, if the top-most method override has a io.smallrye.common.annotation.NonBlocking annotation.
1. blocking, if the class that with the top-most method override has a io.smallrye.common.annotation.Blocking annotation.
1. non-blocking, if the class that with the top-most method override has a io.smallrye.common.annotation.NonBlocking annotation.
1. blocking, if top-most method override has a Transaction annotation.
1. blocking, if the service class or any of its base classes has a Transaction annotation.
1. Else: non-blocking.

Duplicate the existing tests `BlockingMethodsTest` and `BlockingMethodsWithMutinyImplTest` with implementation variants, so each tests these scenarios:
* `@GrpcService` implementation that directly extends the `*ImplBase` class
* `@GrpcService` implementation that extends another class that extends the `*ImplBase` class, `@Blocking` methods on the "top-most" class
* `@GrpcService` implementation that extends another class that extends another class that extends the `*ImplBase` class, `@Blocking` annotation on an "intermediate" base class.

Fixes https://github.com/quarkusio/quarkus/issues/23911 + https://github.com/quarkusio/quarkus/issues/23912

snazy committed:
Handle gRPC `@Blocking` in superclasses as well

Updates the "blocking methods" discovery mechanism in `io.quarkus.grpc.deployment.GrpcServerProcessor`.

Two changes:
* Discover all gRPC methods from the base class that implements `io.grpc.BindableService` and use the names of the discovered methods to find blocking methods.
* Discover whether gRPC methods shall be blocking by inspecting the `@Blocking`/`@NonBlocking`/`@Transactional` annotations on method and class level including base classes.

Whether a gRPC method is blocking is determined using the following rules, first matching rule applies:
1. blocking, if the top-most method override has a io.smallrye.common.annotation.Blocking annotation.
1. not-blocking, if the top-most method override has a io.smallrye.common.annotation.NonBlocking annotation.
1. blocking, if the class that with the top-most method override has a io.smallrye.common.annotation.Blocking annotation.
1. non-blocking, if the class that with the top-most method override has a io.smallrye.common.annotation.NonBlocking annotation.
1. blocking, if top-most method override has a Transaction annotation.
1. blocking, if the service class or any of its base classes has a Transaction annotation.
1. Else: non-blocking.

Duplicate the existing tests `BlockingMethodsTest` and `BlockingMethodsWithMutinyImplTest` with implementation variants, so each tests these scenarios:
* `@GrpcService` implementation that directly extends the `*ImplBase` class
* `@GrpcService` implementation that extends another class that extends the `*ImplBase` class, `@Blocking` methods on the "top-most" class
* `@GrpcService` implementation that extends another class that extends another class that extends the `*ImplBase` class, `@Blocking` annotation on an "intermediate" base class.

Fixes #23911 + #23912

michalszynkiewicz commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+                .collect(Collectors.toSet());
+    }
+
+    private Set<String> gatherBlockingMethods(ClassInfo service, IndexView index) {
```
what if a class has two methods with the same name?

michalszynkiewicz replied:
hmm... I see we are registering blocking methods by name currently too

snazy replied:
Yea - duplicate names aren't allowed in protobuf. I thought, it's easier to read with that change to `String`.

michalszynkiewicz reviewed (commented):
a few notes, no issues; I still need to take a look at the tests

snazy commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+                .flatMap(mi -> mi.annotations().stream())
+                .map(AnnotationInstance::name)
+                .filter(relevant::contains)
+                .collect(Collectors.toSet());
```
Not sure whether I understand correctly. But it's an important aspect.

The idea here is to merge all `@Blocking`, `@NonBlocking` and `@Transactional` annotations from the whole hierarchy. But that might not be what a user would expect.

Are you proposing to only use the annotations on the "top-most" declaration - i.e. the actual service implementation (having the `@GrpcService` annotation) and for methods the annotations from the "top-most" implementation/override?
Not sure how to exactly implement it here though - if we'd have `@Blocking`/`@NonBlocking` only on methods, it can certainly use only the annotations on the "top-most" method implementation/override. It becomes tricky if a method's explicitly declared e.g. `@NonBlocking` in a base class but the actual service implementation class has a `@Blocking` annotation. I tend to use the annotation from the method, as it's more specific.

For example:
```
@GrpcService @NonBlocking
class FooService extend BaseFooService {
  // bar() would be Blocking here
}

class BaseFooService extends FooImplBase {
  @Override
  @Blocking
  void bar() {}
}
```

```
@GrpcService @NonBlocking
class FooService extend BaseFooService {
  // bar() would be NonBlocking here
}

@Blocking
class BaseFooService extends FooImplBase {
  @Override
  void bar() {}
}
```

michalszynkiewicz replied:
hmm... Maybe, for methods, to avoid confusion we should only apply blocking/nonblocking on the actual implementation, so if a method is inherited and not overriden, use the annotation from it, if it is overriden, apply only the annotation from this method (if there is one, none if there's none)?

The class level annotation could still be inheritted (the "closest" one to the implementation).

@mkouba WDYT?

snazy replied:
I've pushed another commit with an updated logic (probably while you've been writing the comment).
The "rule" when a method is considered blocking is described here https://github.com/quarkusio/quarkus/pull/23927/commits/ae3cf7b8074f4aa1656cbdd3e74dd15c6bdf95bb#diff-950bac0249018afb3983346929ea669c55e8d151de6161dc085a39b842ddd5c1R361

snazy replied:
Note: the behavior for existing implementations doesn't change - but I hope it's more intuitive when there are overridden methods.

snazy commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+
+    private Set<String> gatherBlockingMethods(ClassInfo service, IndexView index) {
+        List<ClassInfo> classes = classHierarchy(service, index);
+        Set<DotName> annotation = relevantAnnotations(classes);
```
done

snazy commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+    }
+
+    private static Set<DotName> relevantAnnotations(List<ClassInfo> classHierarchy, MethodInfo method) {
+        Set<DotName> relevant = new HashSet<>();
```
done

snazy commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+
+        // Collect all gRPC methods from the *ImplBase class
+        Map<String, Type[]> grpcMethods = classes.get(classes.size() - 1).methods().stream()
+                .filter(m -> !skipMethod(m))
```
done

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -109,13 +110,13 @@ void processGeneratedBeans(CombinedIndexBuildItem index, BuildProducer<Annotatio
         // 1. there is a user-defined bean that implements the generated interface (injected delegate)
         // 2. there is no user-defined bean that extends the relevant impl bases (both mutiny and regular)
-        for (ClassInfo generatedBean : index.getIndex().getKnownDirectImplementors(GrpcDotNames.MUTINY_BEAN)) {
+        for (ClassInfo generatedBean : index.getIndex().getAllKnownImplementors(GrpcDotNames.MUTINY_BEAN)) {
```
I don't think that this change makes sense. This build step should only process all direct implementors of `MutinyBean`, i.e. the generated beans.

snazy replied:
Think it makes sense to allow inheritance and have multiple implementations (`@Default` and such).

Something like this:
```
@GrpcService
@Default
class MyServiceForProd extends MyServiceBaseImpl {
  protected doSomeWork(...) {
    // ...
  }
}

@GrpcService
@Dependent
@SomeAnnotationForSelection(DeploymentType.DEV)
class MyServiceForDev extends MyServiceBaseImpl {
  protected doSomeWork(...) {
    // ...
  }
}

class MyServiceBaseImpl extends MyServiceBaseImpl {
  @Blocking
  public Uni<...> foo(SomeRequest req) {
    // common code
    getBackend().doSomeWork(...);
    // common code
  }

  protected abstract doSomeWork(...);
}
```

mkouba replied:
But `index.getIndex().getAllKnownImplementors(GrpcDotNames.MUTINY_BEAN)` would not return any of these classes. It should return the beans generated by quarkus.

snazy replied:
Ohhh!! I see.
Weird, the test cases work as expected though.

snazy replied:
Ah - got it now. Yea - that change doesn't make sense.

snazy commented:
@mkouba @michalszynkiewicz is there anything missing for this PR?

mkouba commented:
> @mkouba @michalszynkiewicz is there anything missing for this PR?

I'm sorry but I didn't have time to review the recent changes yet. I will add my review today!

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+     */
+    private static List<ClassInfo> classHierarchy(ClassInfo service, IndexView index) {
+        List<ClassInfo> collected = new ArrayList<>();
+        for (ClassInfo ci = service; ci != null;) {
```
I'd prefer something like 
```java
while (service != null) {
  ...
  service = index.getClassByName(superName);
}
```
It's IMO a bit more readable...

snazy replied:
Updated

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
-                    + "' contains both @Blocking and @NonBlocking annotations.");
+    /**
+     * Generate list of {@link ClassInfo} with {@code service} as the first element and the class implementing
+     * {@code io.grpc.BindableService} (the protobuf generated {@code *ImplBase}) as the last one.
```
Note that [user beans that implement service interface](https://quarkus.io/guides/grpc-service-implementation#implementing-a-service-with-the-mutiny-api) do not have to implement `BindableService` at all.

snazy replied:
Updated the javadoc + inline comment

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
-            Set<MethodInfo> blockingMethods = gatherBlockingMethods(service);
-            for (MethodInfo method : blockingMethods) {
-                item.registerBlockingMethod(method.name());
+            Set<String> blockingMethods = gatherBlockingMethods(service, index);
```
It might make sense to rename `gatherBlockingMethods()` to `gatherBlockingMethodNames()` to make it more clear...

snazy replied:
Done

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+
+        Set<String> result = new HashSet<>();
+
+        // Collect all gRPC methods from the *ImplBase class
```
Again, an `*ImplBase` class does not have to be present.

snazy replied:
Updated the comment

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+        UNDEFINED(false, false),
+        BLOCKING(true, true),
+        NON_BLOCKING(false, true),
+        TRANSACTIONAL(true, false);
```
I wonder if we should use a different name; e.g. `IMPLICIT`, so that it's possible to reuse the constant for multiple annotations (and not only `@Transactional`).

snazy replied:
yup

mkouba commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
-                } else {
-                    result.add(method);
+    /**
+     * Retrieve the blocking-mode for the given method (by name + parameter types).
```
@michalszynkiewicz I thought that gRPC method names must be unique?

So why the usage of param types here?

mkouba reviewed (commented):
Looks good. I've added some comments. Could you squash the commits pls?

snazy commented:
Rebased, addressed comments, squashed

michalszynkiewicz commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -212,41 +213,193 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+     * <li>blocking, if any of the method overrides has a {@link javax.transaction.Transaction} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link javax.transaction.Transaction}
+     * annotation.</li>
+     * <li>Else: non-blocking.</li>
```
As I wrote before, I think this is how it should work:
- apply method-level @Blocking, @NonBlocking and blocking coming from @Transactional only to methods that define it (and not on overriding methods)
- apply class-level annotations only to methods in the class they are in

Even with this simplification, we're ending up with quite strange cases like:
```java
@Blocking
class BaseClass {
   void foo(){}
}

@NonBlocking
class SubClass extends BaseClass {
 @Override
 void foo() {
 }
}


class SubClass2 extends BaseClass {
 @Override
 void foo() {
 }
}
```

or 
```java
class BaseClass {
  @Blocking
   void foo(){}
}

@NonBlocking
class SubClass extends BaseClass {
}
```

michalszynkiewicz replied:
To my knowledge we don't support inheriting `@Blocking` in RestEasy Reactive (yet?).

@Ladicek, maybe you have some opinion on how this would work best?

Ladicek replied:
I certainly have strong opinions on annotation inheritance (it's evil!), but I also know how annotations are inherited in the Java language: they are never inherited from methods, only from superclasses. The Common Annotations specification demands the same, and it's also how CDI behaves. Interestingly, JAX-RS behaves differently, which is probably the source of inspiration here.

Here's my take: the `@Blocking` and `@NonBlocking` annotations are _not_ declared as `@Inherited`. Don't inherit them. Anywhere. Ever.

snazy replied:
Yea, it's been inspired by JAX-RS. But I get the (strong) opinion and agree that this implementation should conform to the documentation of `@Blocking` and `@NonBlocking`. `@Transactional` however is `@Inherited`, which makes it a bit tricky.

Anyway, I've updated the implementation and added a unit test for a bunch of combinations to check the blocking methods for a particular class.

michalszynkiewicz replied:
is there any test that would check if in the following cases the `foo` method is non-blocking?
```java
@Blocking 
class BaseClass { 
   void foo(){} 
} 
class SubClass extends BaseClass { @Override void foo() { } 
} 
```

```java
class BaseClass { 
   @Blocking 
   void foo(){} 
} 
class SubClass extends BaseClass { 
   @Override void foo() { } 
} 
```

michalszynkiewicz replied:
also, what's the reason for including every kind of communication type (unary, client and server side streaming, bidirectional) in the tests for this PR? Isn't it enough to test only one case, e.g. unary?

snazy replied:
I followed what the existing tests were doing to check the blocking behavior so I'm not sure actually.

Re unit test cases: added a couple more following the pattern you asked for.

michalszynkiewicz commented on extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:
```diff
@@ -213,41 +214,171 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+     * <li>non-blocking, if the class that with the top-most method override has a
+     * {@link io.smallrye.common.annotation.NonBlocking} annotation.</li>
+     * <li>blocking, if top-most method override has a {@link javax.transaction.Transaction} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link javax.transaction.Transaction}
```
links should be to `@Transactional`

michalszynkiewicz commented:
FYI, I think what the PR does now is okay.
I planned to simplify the tests and merge it last week but didn't have time, I should be able to do it next week.

michalszynkiewicz committed:
Blocking inheritance for gRPC: simplified tests

michalszynkiewicz commented on extensions/grpc/deployment/src/test/java/io/quarkus/grpc/server/blocking/inheritance/BlockingFromTransactionalTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    @Timeout(5)
+    void shouldBlockOnOverridenTransactional() {
```
this test fails, so it seems the sixth point of the description does not work, i.e. `blocking, if the service class or any of its base classes has a Transaction annotation.`
Or am I missing something?

michalszynkiewicz commented:
@snazy I simplified the tests a bit and I think I found one situation where it doesn't work as described, pelase take a look at  the comment in the test.

I apologize that it took me so long

snazy committed:
Fix case when method's overriden but `@Transactional`'s defined on a superclass

snazy commented:
@michalszynkiewicz the refactoring looks much better! Thanks!

There was a bug when a method's been overridden and `@Transactional` defined on a superclass. Fixed that and also added a test case to `GrpcServerProcessorTest`
